Some cleaning up of the HttpClientTracer hierarchy

trask opened the pull request:
~Motivated by #1429, and by getting rid of `shouldStartSpan`.~

~This PR tackles `HttpClientTracer` subclasses, to get a reasonably broad view of whether this is a good idea or not.~

~Primarily, it converts `HttpClientTracer.startSpan()` to `HttpClientTracer.startOperation()`, returning a new concept Operation, instead of returning Context.~

~This Operation has the `end()` methods on it now, so that it can end spans, metrics, or do nothing if it's a no-op.~

~This eliminates the need for having to call the `shouldStartSpan()` methods before calling `startSpan()`.~

~And it generally reduces the amount of code in advice classes (e.g. way less `Java8BytecodeBridge`).~

~Btw, I think I like how it's the "Operation" that is now propagated via the library-specific attributes for the purpose of calling end somewhere else, just because it's a bit more concrete than "oh, we're propagating yet another Context".~

See the new scope of this PR outlined in [comment below](https://github.com/open-telemetry/opentelemetry-java-instrumentation/pull/1840#issuecomment-743718165).

Closes #1009

trask commented on instrumentation/http-url-connection/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/httpurlconnection/DefaultHttpUrlConnectionOperation.java:
```diff
@@ -0,0 +1,38 @@
+
+  @Override
+  public boolean finished() {
+    return finished;
```
oh, yes!

trask commented on instrumentation/netty/netty-3.8/javaagent/src/main/java/io/opentelemetry/javaagent/instrumentation/netty/v3_8/client/NettyHttpClientTracer.java:
```diff
@@ -25,6 +34,36 @@ public static NettyHttpClientTracer tracer() {
+      ChannelHandlerContext ctx, MessageEvent msg, ChannelTraceContext channelTraceContext) {
+
+    if (!(msg.getMessage() instanceof HttpRequest)) {
+      ctx.sendDownstream(msg);
```
oh that was a very bad copy-paste, thx

trask commented on instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/DefaultHttpClientOperation.java:
```diff
@@ -0,0 +1,71 @@
+  private final Context context;
+  // TODO separate implementation when parentContext is not needed (memory optimization)?
+  private final Context parentContext;
+  // TODO (trask) move end/endExceptionally/etc to here and remove tracer reference
+  private final HttpClientTracer<?, ?, RESPONSE> tracer;
```
I'm going to work a bit more on this PR. I'll move it back to draft status for now.

I think it will be useful to see how it looks once we start moving things over here from tracers, e.g.
* do tracers just become Operation factories, exposing a single public method `startOperation`?
* can we strictly limit the public surface of `HttpClientOperation`, so that we don't need a separate no-op implementation for each subclass

iNikem replied:
I took only a cursory look at this and I think I like it. But why "moving things over here from tracers" is a good direction? Why is an `Operation` a better place for a code than a `Tracer`? Because it is not limited to tracing?

trask commented:
Some notes, now that I have better idea of how this is shaping up:
* Operation is not intended for subclassing for instrumentation-specific behavior
  * (though abstract Tracers that are part of `instrumentation-api` will likely have their own `Operation` implementations)
* Tracer is the place for instrumentation-specific behavior, but it only has one public method `startOperation(...)` and the rest of the instrumentation-specific behavior is implemented via protected method overrides
  * the rest of the communication from advice to Tracers is done via the public methods on Operations which call the methods in the abstract Tracer class (which they have access to since they are in same package)

iNikem commented:
I will probably make a lot of comments about specific code parts later, but let me start with "architectural discussion".

We are introducing new stateful "interface" called `Operation`, which mostly models the lifecycle of a `Span` that it wraps. This allows us to have `noop` operation and thus eliminate the need to call `shouldStartSpan` from all advices. And just call `operation.end()` without null-checking the `scope`.

`Tracers` continue to encapsulate semantic conventions. Operations now will communicate with Tracers instead of Advices, which allows us to reduce public surface of Tracers. I assume that in the future we will have Meters as well and Operations, not Advices, will talk to them too.

Tracers continue to serve as extension points. If somebody wants to alter attributes or modify produced telemetry, they are expected to extend Tracers, not Operations.

Is my understanding of the intent of this change correct?

If yes, then I have problems with `LazyHttpClientOperation` :) I try to understand which method should live where. If Tracers are for semantic convention and Operations are for lifecycle, then `LazyHttpClientOperation.onRequest()` is strange. How do we decide which methods should go to Operation and which to Tracer? Is it Ok for `Operation.onRequest` or similar method to add attributes to span or they must delegate to tracer? This seems like we have keep two classes in sync.

trask commented:
> Is my understanding of the intent of this change correct?

:thumbsup:

> How do we decide which methods should go to Operation and which to Tracer? Is it Ok for Operation.onRequest or similar method to add attributes to span or they must delegate to tracer? This seems like we have keep two classes in sync.

I'll try moving everything to Operation to see what that looks like, I think that might make things clearer.

iNikem commented:
>moving everything to Operation

You mean including semantic convention related stuff?

trask commented:
> > moving everything to Operation
> 
> You mean including semantic convention related stuff?

moving everything to Operation didn't work out well, so went in the other direction and moved those methods `end`, etc back into Tracer, and those methods now accept `Operation` instead of `Context`

trask reviewed (commented):
ready for review

iNikem commented:
> > > moving everything to Operation
> > 
> > 
> > You mean including semantic convention related stuff?
> 
> moving everything to Operation didn't work out well, so went in the other direction and moved those methods `end`, etc back into Tracer, and those methods now accept `Operation` instead of `Context`

Wait, if `end` is back in Tracer, then what is the purpose of Operation? And how this solves the initial problem of not ending wrong spans? Can you please provide a description of how the current state works?

trask commented:
> And how this solves the initial problem of not ending wrong spans?

If you only have Context, then you don't know if the Span in the Context should be ended or not.

If you have Operation, you can have no-op version, which returns no-op Span, or you can have real version, which returns real Span.

mateuszrzeszutek commented:
> moving everything to Operation didn't work out well, so went in the other direction and moved those methods end, etc back into Tracer, and those methods now accept Operation instead of Context

Just bring curious: what exactly didn't work out? On the paper having `end()` methods in `Operation` seemed like a good idea,

trask commented:
Moved the public `end()` methods back to `Operation`, while keeping all extension points inside of the Tracer.

Pros
* Feels natural to call end directly on the operation, instead of having to pass the operation to the tracer

Cons
* `Operation` class needs to be parameterized on `RESPONSE`
* `Operation` needs reference to the tracer
* Probably will want two Operation interfaces, one for operations that have a response object and one for those that don't

trask commented:
Moving back to Draft.

Discussed with @anuraaga and @bogdandrutu in SIG meeting. Going to try to eliminate the Operation concept, and add a `DelegatingSpan` to the Context in case of suppressed (e.g. nested CLIENT) Span. The `DelegatingSpan` will delegate everything to parent Span except for `end()`.

I'm not fully convinced this will work, because the suppressed instrumentation will end up setting a bunch of attributes on the parent span that may or may not make sense, but only one way to find out for sure, and maybe we'll see some other options emerge to deal with that problem once I prototype.

trask commented on instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/context/ParentPropagatingContext.java:
```diff
@@ -0,0 +1,42 @@
+import io.opentelemetry.context.ContextKey;
+import io.opentelemetry.context.Scope;
+
+public class ParentPropagatingContext implements Context {
```
this is what tracers can return from `startOperation` if they want to suppress the span

trask reviewed (commented):
`startOperation()` now always takes `parentContext` as the first arg, and returns `Context`, which is nice

still nervous about passing attributes from the suppressed child through to the parent, see `ParentPropagatingContext` as alternative to avoid that for now at least

trask commented:
Since no-ops have to perform some memory allocation with this approach (as opposed to `Operation.noop()`), it's not ideal for no-op'ing the instrumented overloaded methods case, and we may want to keep CallDepth optimization for those

trask commented:
maybe `shouldStartSpan()` is not such a bad idea, to avoid extra allocation in both op and no-op cases 😭

trask commented:
Renamed this PR to "Some cleaning up of the HttpClientTracer hierarchy".

It now tackles a lot less, keeping in place the existing `shouldStartOperation` for now, and we can still try other approaches to get rid of `shouldStartOperation` in the future.

Summary of what's in this PR:
* Rename shouldStartSpan to shouldStartOperation, and use it consistently everywhere
* Rename startSpan to startOperation
* Always pass in the parentContext to startOperation and return the new context
* Moved code from Advice into Tracers where it made sense
* Other related cleanup of the http client instrumentation and tracers

Please (re-)review, then we can move forward with #1871 from @parallelstream.

iNikem commented on instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/context/ContextWithParent.java:
```diff
@@ -0,0 +1,27 @@
+
+import io.opentelemetry.context.Context;
+
+public class ContextWithParent {
```
Doesn't this look like we need `Context.getParent()` method? @anuraaga ?

anuraaga replied:
Don't think so, there's no concept of parenting in `Context`, that's a tracing concept. e.g., `context.with(animal, "cat").with(car, "nissan")`, the intermediate contexts naturally aren't parents, but we have no way of knowing what the "logical parent" is at the context layer.

It's one reason how I've generally encouraged the word `invocationContext` instead of parent, but haven't gotten that to stick :P We need to keep a reference to the invocation context to remount it in asynchronous callbacks. 

Normally we would not need this sort of class since, e.g., both contexts would just get scoped into a lambda callback for an async handler. It seems here we're using it for `InstrumentationContext` - I suspect the reason we can't store them as two fields is because we only support a max of one fields for any given type. If this reasoning makes sense, we may want to move this class to the javaagent API instead of exposing it to library instrumentation which shouldn't really need it and at a glance it seems like a class specific to our javaagent needs.

iNikem commented on instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java:
```diff
@@ -67,6 +67,16 @@ public Span startSpan(String spanName, Kind kind) {
   }
 
+  // TODO (trask) make the key private
+  protected final boolean inClientSpan(Context parentContext) {
```
Why `protected final`? It looks like default visibility to me.

iNikem commented on instrumentation-api/src/main/java/io/opentelemetry/instrumentation/api/tracer/BaseTracer.java:
```diff
@@ -140,16 +140,14 @@ public void endExceptionally(Span span, Throwable throwable) {
-    addThrowable(span, throwable);
-  }
-
-  protected Throwable unwrapThrowable(Throwable throwable) {
```
I am not sure this is a good idea

iNikem reviewed (commented):
Sorry @trask , but can I ask you to split this PR? It is huge and it does contain independent changes. Some of the them are mechanical, like method rename, and some are not. It is not easy to review that in one go.

trask commented:
sure, will split it up tomorrow

anuraaga commented on instrumentation/aws-sdk/aws-sdk-2.2/library/src/main/java/io/opentelemetry/instrumentation/awssdk/v2_2/AwsSdk.java:
```diff
@@ -40,29 +38,21 @@ public static Tracer tracer() {
+   * Returns the {@link Operation} stored in the {@link ExecutionAttributes}, or {@code null} if
+   * there is no operation set.
    */
-  public static ExecutionInterceptor newInterceptor(Kind kind) {
```
Yup, thanks

trask commented:
I split out the more complex parts into #1907, #1908, #1909, #1910, #1911. Will do the more mechanical parts after.

iNikem commented:
Should we move this back to Draft and invalidate all reviews?

trask commented:
Closing, probably nothing in here that is still relevant 👍
